File: dotnet/src/SemanticKernel/TemplateEngine/Blocks/Block.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Microsoft.SemanticKernel.Orchestration;

namespace Microsoft.SemanticKernel.TemplateEngine.Blocks;

/// <summary>
/// Base class for blocks parsed from a prompt template
/// </summary>
public abstract class Block
{
    internal virtual BlockTypes Type => BlockTypes.Undefined;

    internal virtual bool? SynchronousRendering => null;

    /// <summary>
    /// The block content
    /// </summary>
    internal string Content { get; }

    /// <summary>
    /// App logger
    /// </summary>
    protected ILogger Log { get; } = NullLogger.Instance;

    /// <summary>
    /// Base constructor
    /// </summary>
    /// <param name="content">Block content</param>
    /// <param name="log">App logger</param>
    protected Block(string? content, ILogger? log = null)
    {
        if (log != null) { this.Log = log; }

        this.Content = content ?? string.Empty;
    }

    internal virtual Task<string> RenderCodeAsync(SKContext executionContext)
    {
        throw new NotImplementedException("This block doesn't support async code execution, use Render(ContextVariables) instead.");
    }

    internal virtual string Render(ContextVariables? variables)
    {
        throw new NotImplementedException("This block requires async code execution, use RenderCodeAsync(SKContext) instead.");
    }

    internal abstract bool IsValid(out string error);
}

File: semantic_kernel/template_engine/blocks/block.py
# Copyright (c) Microsoft. All rights reserved.

from abc import ABC, abstractmethod
from logging import Logger
from typing import Optional, Tuple

from semantic_kernel.orchestration.context_variables import ContextVariables
from semantic_kernel.orchestration.sk_context import SKContext
from semantic_kernel.template_engine.blocks.block_types import BlockTypes


class Block(ABC):
    _type: BlockTypes
    _content: str
    _log: Logger

    def __init__(self, block_type: BlockTypes, content: str, log: Logger) -> None:
        self._type = block_type
        self._content = content
        self._log = log

    async def render_code_async(self, context: SKContext) -> str:
        raise NotImplementedError("This block does not support code execution")

    @abstractmethod
    def is_valid(self) -> Tuple[bool, str]:
        pass

    @abstractmethod
    def render(self, variables: Optional[ContextVariables]) -> str:
        pass

    @property
    def type(self) -> BlockTypes:
        return self._type

    @property
    def content(self) -> str:
        return self._content
